src/compiler/checker.ts

```diff
@@ -5667,6 +5752,191 @@ namespace ts {
+            for (const member of (node as TypeLiteralNode).members) {
+                if (member.kind === SyntaxKind.SpreadTypeElement) {
+                    if (members) {
+                        spreads.push(createAnonymousType(node.symbol, members, emptyArray, emptyArray, stringIndexInfo, numberIndexInfo));
+                        members = undefined;
+                        stringIndexInfo = undefined;
+                        numberIndexInfo = undefined;
+                    }
+                    spreads.push(getTypeFromTypeNode((member as SpreadTypeElement).type));
+                }
+                else if (member.kind === SyntaxKind.IndexSignature) {
+                    const index = member as IndexSignatureDeclaration;
+                    if (index.parameters.length === 1) {
+                        const parameter = index.parameters[0];
+                        if (parameter && parameter.type) {
+                            const indexInfo = createIndexInfo(index.type ? getTypeFromTypeNode(index.type) : anyType,
+                                                              (getModifierFlags(index) & ModifierFlags.Readonly) !== 0, index);
+                            if (parameter.type.kind === SyntaxKind.StringKeyword) {
+                                stringIndexInfo = indexInfo;
+                            }
+                            else {
+                                numberIndexInfo = indexInfo;
+                            }
+                        }
+                    }
+                }
+                else if (member.kind !== SyntaxKind.CallSignature && member.kind !== SyntaxKind.ConstructSignature) {
+                    // note that spread types don't include call and construct signatures
+                    const flags = SymbolFlags.Property | SymbolFlags.Transient | (member.questionToken ? SymbolFlags.Optional : 0);
+                    const text = getTextOfPropertyName(member.name);
+                    const symbol = <TransientSymbol>createSymbol(flags, text);
+                    symbol.declarations = [member];
+                    symbol.valueDeclaration = member;
+                    symbol.type = getTypeFromTypeNodeNoAlias((member as IndexSignatureDeclaration | PropertySignature | MethodSignature).type);
+                    if (!members) {
+                        members = createMap<Symbol>();
+                    }
+                    members[symbol.name] = symbol;
+                }
+            }
+            if (members || stringIndexInfo || numberIndexInfo) {
+                spreads.push(createAnonymousType(node.symbol, members || emptySymbols, emptyArray, emptyArray, stringIndexInfo, numberIndexInfo));
+            }
+            return getSpreadType(spreads, node.symbol, aliasSymbol, aliasTypeArguments);
+        }
+
+        function getSpreadType(types: Type[], symbol: Symbol, aliasSymbol?: Symbol, aliasTypeArguments?: Type[]): Type {
+            if (types.length === 0) {

```

Comment:
It's easier to do it inside `checkObjectLiteral` and `getTypeFromTypeLiteral` because those bodies of code both have a for loop and would otherwise need to build up a list to hold the spread types. This way they can just keep a single spread type and manually fold in new spreads when needed, like `spread = getSpreadType(spread, newSpread)`